pkg/backends/calico/routes.go

```diff
@@ -267,21 +349,56 @@ func (rg *routeGenerator) unsetRouteForSvc(obj interface{}) {
 	defer rg.Unlock()
 
 	// Remove the route if it exists.
-	if route, exists := rg.svcRouteMap[key]; exists {
-		rg.withdrawRoute(key, route)
+	if route, exists := rg.svcClusterRouteMap[key]; exists {
+		rg.withdrawClusterRoute(key, route)
 	}
+
+	// Remove all External IP's
+	rg.unsetExternalRoutesForSvc(key)
+
 }
 
-// advertiseRoute advertises a route and caches it.
-func (rg *routeGenerator) advertiseRoute(key, route string) {
-	rg.svcRouteMap[key] = route
+// unsetExternalRoutesForSvc withdraws all routes for the service with
+// the given key.
+func (rg *routeGenerator) unsetExternalRoutesForSvc(key string) {
+	if rg.svcExternalRouteMap[key] != nil {
+		for route := range rg.svcExternalRouteMap[key] {
+			rg.withdrawExternalRoute(key, route)
+		}
+	}
+}
+
+// advertiseClusterRoute advertises a route for a service ClusterIP and caches it.
+func (rg *routeGenerator) advertiseClusterRoute(key, route string) {
+	rg.svcClusterRouteMap[key] = route
 	rg.client.AddStaticRoutes([]string{route})
 }
 
-// withdrawRoute withdraws a route and removes it from the cache.
-func (rg *routeGenerator) withdrawRoute(key, route string) {
+// withdrawClusterRoute withdraws a route and removes it from the ClusterIP cache.
+func (rg *routeGenerator) withdrawClusterRoute(key, route string) {
 	rg.client.DeleteStaticRoutes([]string{route})
-	delete(rg.svcRouteMap, key)
+	delete(rg.svcClusterRouteMap, key)
+}
+
+// advertiseExternalRoute advertises a route for a service ExternalIP and
+// caches it.
+func (rg *routeGenerator) advertiseExternalRoute(key, route string) {
```

Comment:
Do you think it would be simpler to have a single map for both cluster IPs and external IPs?

e.g.

```
if advertise {
  routes := getAllRoutesForService(svc, endpoints)
  advertiseRoutes(key, routes)
} else {
  routes = previouslyAdvertised(svc)
  withdrawRoutes(key, routes)
}
```

`getAllRoutesForService` would just take the clusterIP, plus any external IPs that are whitelisted and return them all.

I like this because it make the logic generic across the IP types, and we don't need to duplicate effort with `withdrawClusterRoute` and `withdrawExternalRoute` (which are logically and functionally very similar).

What do you think?

Reply:
I agree, this would be more elegant.

Reply:
@tompntn are you alright making that change?